client.go

```diff
@@ -604,14 +604,25 @@ func (c *client) startCommsWorkers(conn net.Conn, connectionUp connCompletedFn,
 	// The connection is now ready for use (we spin up a few go routines below). It is possible that
 	// Disconnect has been called in the interim...
 	if err := connectionUp(true); err != nil {
-		DEBUG.Println(CLI, err)
-		close(c.stop) // Tidy up anything we have already started
-		close(incomingPubChan)
-		c.workers.Wait()
-		c.conn.Close()
-		c.conn = nil
-		return false
+		ERROR.Println(CLI, err)
+
+		/* issue 675：goroutine leak when connectionUp(true) return error
+		 *   Only when status == disconnecting will this logic be executed.
+		 * The goroutine that changes the status to disconnecting should be
+		 * responsible for resource cleanup (which is indeed how it is done).
+		 *
+		 *   Being disconnected right when the connection is established is a special case.
+		 * Even if we remove this check for connectionUp(true), the program must still function correctly,
+		 * as if a Disconnect event occurred immediately after connectionUp(true) completed.
+		 */
+
+		//close(c.stop) // Tidy up anything we have already started
```

Comment:
So, in summary, this PR will change things so that:

 if a connection is in progress when `Disconnect` is called then the connection attempt will be completed (including calling `OnConnect` and spinning up goroutines) and then the `Disconnect` operation will proceed after this (so, effectively a normal disconnect). In this case `connectionStatus.Disconnecting()` will wait until the connection process has completed before returning (and allowing the disconnection process to proceed). `startCommsWorkers` maintains a lock on `connMu` throughout the process, so `stopCommsWorkers` will be blocked until it completes.

My feeling is that this should work (but it's quite difficult to trace all possibilities). However I'd prefer to see the old code removed rather than being commented out (it will be available in the git history) and would like to see a comment something like the above (i.e. we will allow the connection to complete before the Disconnect is allowed to proceed).

I think I'm OK to accept this but if you are able to add something like the following I'd appreciate it. Adding a test for this would be absolutely fantastic (but I'm thinking it might be quite difficult to write).

Reply:
Of course, I have followed your suggestions. 
For testing, I simulated this scenario locally using the method I previously used to debug the issue, and it works.

Reply:
Thanks.

>For testing, I simulated this scenario locally using the method I previously used to debug the issue, and it works.

This comment was more about adding a test (that does not pass with the old code but does with the new).  The main purpose of this is to help prevent regressions (the state management stuff is pretty complex and Disconnect is not used all that often so it would be east to reintroduce the issue). Anyway I'll accept this PR now and if you do have a test that could be added later.